### PR TITLE
ZEN-423: Pending and Modified Button States

### PR DESCRIPTION
**Pull request by mpcarolin**
**Ticket**: [ZEN-423](https://jira.simpleviewtools.com/browse/ZEN-423)

Associated keg-hub PR: https://github.com/simpleviewinc/keg-hub/pull/63

## Context

* currently, our booking buttons (both group-booking and single) do not update based on the status of their booking requests
* they need to show a loading spinner, until the request completes

## Goal

* For group booking and single booking, show a loading spinner until the booking request completes
* update the UI to look good around this
* also disable modal booking buttons until modifications are made

## Updates
* `apps/Sessions.js`
  * removed reference to EFDemoContainer since we do not need it any longer
* `src/actions/session/booking/sessionBookingRequest.js` && `src/actions/session/booking/sessionWaitingListRequest.js`
  * updated to store a pending session object in the items store
  * also emit and event for the "pending" state transitions
* `src/actions/session/booking/setUserModifiedBooking.js`
  * added action to track when the group booking modal is modified
  * this is needed to ensure users cannot select "book" until they have modified the existing attendees
* `src/actions/session/mapSessionInterface.js`
  * added logic for updating the pending state based on incoming props 
* `src/actions/session/pending/clearPendingSession.js` && `.../setPendingSession`
  * created actions for setting and removing the pending session
* `src/components/booking/attendeeBookingList.js`
  * updated list to be disabled if the group booking session is pending 
* `src/components/booking/groupBooker.js`
  * added ui for button spinner, as well as logic for handling the pending state
  * dismisses modal when pending state transitions to not-pending
* `src/components/button/bookingButton.js`
  * updated with `isProcessing` styles when the associated session is pending
* `src/components/button/evfButton.js`
  * updated to allow processing styles to be defined by consumer
* `src/components/loading/evfLoading.js`
  * updated to allow consumer to set color
* `src/components/testData/testData.js`
  * updated with booking delay dropdown ui
* `src/containers/rootContainer.js`
  * updated to simulate a consumer booking-request update 
* `src/hooks/booking/useBookingLists.js`
  * refactored
  * updated hook to also return the *existing* lists (as reflected by the props)
  * updated to ensure it does NOT preselect attendees if at least one attendee is on the booking list (quick fix to a reported bug)
* `src/hooks/booking/useBookingState.js`
  * updated to include the `pending` state of a session with the booking button state
* `src/hooks/booking/useInitGroupBooking.js`
  * updated to start monitoring the booking and waiting lists for modification,
  * if modified, it will flag the session as "modified" so that the user can select the booking button (disabled when no modifications are yet made)
* `src/hooks/booking/useSessionBooking.js`
  * renamed some variables to be more clear
* `src/models/session/pendingSession.js`
  * added pending session model
* `src/reducers/initialStates/**.js`
  * added new initial states for new state I'm tracking:
    * pending session
    * modified session
* `src/utils/booking/bookRequestCompleted.js` && `.../waitRequestCompleted`
  * added utility that compares the pending lists against the new attendees list,
    to determine if the respective requests have completed
* `src/utils/models/sessions/bookingStateFactory.js`
  * updated the `getDisabled` function to set buttons as disabled if there is a single pending session. They become enabled again once that pending session is no longer pending.
* `src/utils/pending/emitPendingEvent.js`
  * helper to emit the pending update event

## Testing

* `keg evf pack run package=docker.pkg.github.com/simpleviewinc/keg-packages/tap:booking-modal-loading`

* New testing features to know about
  * when you submit a booking or wait-list request, that request will now actually
    update the state of the app (only when testing, though, not in the built export)
    so that the attendees are put onto or removed from the right booking and wait lists
  * you can pass a `bookDelay` url parameter (e.g. local.kegdev.xyz/?bookDelay=1), or select the `Booking Delay` dropdown in the testing drawer, to change the amount of time (in seconds) that the app waits before rerendering with the updated attendee state
    * -1 url parameter instructs the app to wait indefinitely

### Modification and Disabled States
* in the default state that enables group booking, select a limited session
  * e.g. `Session on day 2 - limited capacity`, the first session shown on day 2
* in the group booking modal that appears, first verify that the `Book Selected` button is **disabled**.
  * this is disabled because the user has not yet made a modification to the booking and waiting lists
  * you should also see a red asterisk next to the `Select sessions for your group:` instructions at the top of the modal, suggesting to the user that a change is required
* now, select any attendee who is not already checked
  * verify that the book-selected button becomes enabled, and that the red asterisk disappears
* try any combination of attendee list changes, and verify that the disabled and asterisk UI always looks and behaves like you would expect
* finally, open a session that is unlimited and has preselected attendees
  * e.g. `session at same time other session with presenters and labels`, the second session on day 1
  * verify that the booking button is not disabled, since these attendees are pre-selected
  * verify that if you uncheck ALL attendees (returning the session to its initial state), the button becomes disabled

### Pending states
* load `local.kegdev.xyz/?bookDelay=-1`
  * this will make the booking requests never complete
* try booking a session
  * verify that the booking button enters a pending state
  * it will show a loading spinner and the text "Processing"
  * verify that OTHER session booking buttons are put into a disabled state at this time
  * verify that dismissing the modal does not cancel the pending/loading state of the button
* try booking a session again, but this time, while its stuck loading, open the test data json editor.
  * edit the "alert" object at the bottom, making sure you set both its title and message
  * apply the changes
  * verify that the booking button stops loading 
* now load `local.kegdev.xyz/?bookDelay=1`
  * this will make the booking request take 1 second before completing
* submit a booking request in a group booking modal. Then verify:
  * it shows the processing text and styles at first
  * then dismisses the modal when completed, and no longer shows the "processing" text and styles
  * other booking buttons are enabled again
  * the state updates to reflect your change
    * for example, when the modal dismisses itself, you should now see that booking button show the number of attendees you selected (or put on the wait list)
  * try this with any combination of changes:
    * adding to and removing from the booking list
    * adding to and removing from the waiting list
* do the same testing for single-booking states (ignoring modal-specific steps)

### Build
* `keg consumer pack run package=docker.pkg.github.com/simpleviewinc/keg-packages/keg-test-consumer:master`
  * verify the build looks/behaves correctly

### CODE REVIEW

File: src/components/button/bookingButton.js [hunk 1]
```diff
@@ -115,12 +116,16 @@ export const BookingButton = props => {
...
 
+  const pendingStyles = useStyle('button.bookingButton.pending')
+
   return (
     (bookingModel?.text && (
       <EvfButton
         type={bookingModel.state}
         onClick={selectSessionCb}
-        disabled={bookingModel.disabled}
+        disabled={bookingModel.pending || bookingModel.disabled}
```
Comment (daniel112) [inline src/components/button/bookingButton.js:126]: you shouldnt need to pass `bookingModel.pending` here if you're passing it to `isProcessing`
if isProcessing is true, the button will be disabled

File: src/actions/session/pending/setPendingSession.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
+import { Values, ActionTypes } from 'SVConstants'
+import { isStr, isArr, isObj, validate } from '@keg-hub/jsutils'
+import { dispatch } from 'SVStore'
+import { emitPendingEvent } from 'SVUtils/pending/emitPendingEvent'
+const { CATEGORIES } = Values
+
+/**
+ * @param {Object} data
```
Comment (daniel112) [inline src/actions/session/pending/setPendingSession.js:8]: looks like the data is suppose to be of pendingSession model right?

File: src/utils/booking/__tests__/bookRequestCompleted.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import { bookRequestCompleted } from '../bookRequestCompleted'
+
+describe('bookRequestCompleted', () => {
+  it('should return true if the incoming attendees matches the pending list', () => {
```
Comment (daniel112) [inline src/utils/booking/__tests__/bookRequestCompleted.js:4]: this case might be one we never run into but i think it's still possible. 
if incomingAttendees includes another attendee that contains the same `sessionId` that we're checking against && that attendee is not in the `pendingList`. i would expect this to return true i think

**Review by daniel112 — CHANGES_REQUESTED** [review]
testing looks good as well as unit tests

some minor stuff i noticed and what we talked about in chat previously

File: src/utils/pending/emitPendingEvent.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { Values } from 'SVConstants'
+import { getEventEmitter } from 'SVUtils/events/event_emitter'
+import { getStore } from 'SVStore'
+const { CATEGORIES, EVENTS } = Values
+
+const eventManager = getEventEmitter()
+
+/**
+ * Notifies pending session event observers of an update
+ * @param {Object} nextPendingSession
```
[the remark this answers is not in the document]
Comment (mpcarolin) [reply src/utils/pending/emitPendingEvent.js:10]: updated

File: src/actions/session/booking/index.js [hunk 5]
```diff
@@ -1,6 +1,7 @@
 export * from './sessionBookingRequest'
```
[the remark this answers is not in the document]
Comment (mpcarolin) [reply src/actions/session/booking/index.js:1]: as discussed, this issue is unrelated to this PR and were on develop

File: src/utils/models/sessions/bookingStateFactory.js [hunk 6]
```diff
@@ -76,9 +76,16 @@ const getBookingStopped = session => {
  * @returns {boolean} - If the display interaction should be disabled
  */
 const getDisabled = (
-  { session, bookableCount, bookingMode, timeConflicts },
+  { session, pendingSession, bookableCount, bookingMode, timeConflicts },
```
[the remark this answers is not in the document]
Comment (mpcarolin) [reply src/utils/models/sessions/bookingStateFactory.js:79]: done

File: src/hooks/booking/useBookingState.js [hunk 7]
```diff
@@ -61,6 +62,7 @@ const useStoreData = () => {
  * @param {Array} bookingLists - List of attendee ids that are booked or are on the waiting list
  * @param {Object} timeConflicts - Key value pairs of attendees booked in conflicting sessions
  * @param {Array} bookableCount - Attendees that can book the current session
+ * @param {Object} pendingSession - the object indicating if a session has submitted a booking request
```
[the remark this answers is not in the document]
Comment (mpcarolin) [reply src/hooks/booking/useBookingState.js:65]: updated

File: src/theme/components/button/evfButton.js [hunk 8]
```diff
@@ -220,7 +222,19 @@ const bookingButtonStates = reduceObj(
   }
 )
 
+const pendingStyles = {
```
[the remark this answers is not in the document]
Comment (mpcarolin) [reply src/theme/components/button/evfButton.js]: updated

File: src/components/button/bookingButton.js (hunk 1 above)
Comment (mpcarolin) [reply src/components/button/bookingButton.js]: updated

**Pull request closed without merging**